bittide-instances/exe/post-vex-riscv-tcp-test/Main.hs

```diff
@@ -80,51 +97,62 @@ case_testGdbProgram = do
+              Nothing -> pure ()
+              Just h -> do
+                putStrLn "Picocom StdErr"
+                readRemainingChars h >>= putStrLn
+
         hSetBuffering picocomStdInHandle LineBuffering
         hSetBuffering picocomStdOutHandle LineBuffering
 
-        putStr "Waiting for Picocom to be ready..."
-        hFlush stdout
-        waitForLine picocomStdOutHandle "Terminal ready"
-        putStrLn " Done"
+        putStrLn "Waiting for Picocom to be ready..."
+        terminalReadyResult <-
+          timeout 10_000_000 $ waitForLine picocomStdOutHandle "Terminal ready"
+        when (isNothing terminalReadyResult) $ do
+          loggingSequence
+          assertFailure "Timeout waiting for \"Terminal ready\""
 
         putStrLn "Starting GDB..."
         withCreateProcess gdbProc $ \_ (fromJust -> gdbStdOut) _ _ -> do
-          -- Wait for GDB to program the FPGA. If successful, we should see
-          -- "going in echo mode" in the picocom output.
           hSetBuffering gdbStdOut LineBuffering
 
-          putStr "Waiting for GDB to program the FPGA..."
-          hFlush stdout
-          waitForLine picocomStdOutHandle "listening"
-          putStrLn " Done"
-
-        putStrLn "Starting TCP Spray..."
-        -- Test TCP echo server
-        withCreateProcess tcpSprayProc $ \_ (fromJust -> tcpStdOut) (fromJust -> tcpStdErr) _ -> do
-          (stdOutContents, stdErrContents) <- do
-            stdOutContents' <- hGetContents' tcpStdOut
-            stdErrContents' <- hGetContents' tcpStdErr
-            pure (stdOutContents', stdErrContents')
-          picocomOut <- hGetContents picocomStdOutHandle
-
-          case maybePicocomStdErr of
-            Nothing -> pure ()
-            Just h -> do
-              putStrLn "Picocom StdErr"
-              readRemainingChars h >>= putStrLn
-
-          putStrLn ""
-          putStrLn "Picocom stdout"
-          putStrLn $ readUntil "listening" picocomOut
-
-          putStrLn ""
-          putStrLn "tcpSpray StdOut"
-          putStrLn stdOutContents
-
-          if not (null stdErrContents)
-            then assertFailure $ "tcpspray failed with stderr:\n" <> stdErrContents
-            else pure ()
+          putStrLn "Waiting for \"Starting TCP Client\""
+          startingClientResult <-
+            timeout 10_000_000 $ waitForLine picocomStdOutHandle "Starting TCP Client"
+          when (isNothing startingClientResult) $ do
+            loggingSequence
+            assertFailure "Timeout waiting for \"Starting TCP Client\""
+
+          let numberOfClients = 1
+          putStrLn $ "Waiting for " <> show numberOfClients <> " clients to connect to TCP server."
+          maybeClients <- timeout 10_000_000 $ waitForClients numberOfClients serverSock
+          clients <- case maybeClients of
+            Just clients' -> pure clients'
+            Nothing -> do
+              loggingSequence
+              assertFailure "Timeout waiting for clients to connect"
```

Comment:
You're doing almost this three times already.
Maybe factor out something like:
```haskell
foo :: Int -> String -> IO a -> IO a
foo n msg f = do
  res <- timemout n f
  case res of
    Just x -> pure x
    Nothing -> do
      loggingSequence
      assertFailure $ "Timeout waiting for" <> msg
```

And then you can do:
```suggestion
          clients <- foo 10_000_000 "clients to connect" (waitForClients numberOfClients serverSock)
```

Note that `foo` should be locally bound within `case_testGdbProgram` so it can access `loggingSequence`.